### PR TITLE
8292891: ifdef-out some CDS-only functions

### PR DESCRIPTION
**Pull request by iklam**
Some CDS functions are always compiled, even when CDS is not enabled (e.g., for the minimal VM).

This RFE puts some of the obvious ones inside `#if INCLUDE_CDS` blocks. 

Note: my goal is not to make the minimal VM as small as possible. But rather, I don't want to put `#if INCLUDE_CDS` inside each of those functions that access a CDS-only feature. (E.g., `ConstantPoolCache::save_for_archive`, which accesses `_initial_entries`, which is declared only when CDS is enabled).

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8292891](https://bugs.openjdk.org/browse/JDK-8292891): ifdef-out some CDS-only functions

### Reviewers
 * [Coleen Phillimore](https://openjdk.org/census#coleenp) (@coleenp - **Reviewer**) ⚠️ Review applies to [1a42c1db](https://git.openjdk.org/jdk/pull/10010/files/1a42c1db3f1774979a094326fe3c262d49493001)
 * [Calvin Cheung](https://openjdk.org/census#ccheung) (@calvinccheung - **Reviewer**) ⚠️ Review applies to [5f8bcd39](https://git.openjdk.org/jdk/pull/10010/files/5f8bcd39b44caa52b0965df6f18b4141f93524c3)
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**) ⚠️ Review applies to [5f8bcd39](https://git.openjdk.org/jdk/pull/10010/files/5f8bcd39b44caa52b0965df6f18b4141f93524c3)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/10010/head:pull/10010` \
`$ git checkout pull/10010`

Update a local copy of the PR: \
`$ git checkout pull/10010` \
`$ git pull https://git.openjdk.org/jdk pull/10010/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 10010`

View PR using the GUI difftool: \
`$ git pr show -t 10010`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/10010.diff">https://git.openjdk.org/jdk/pull/10010.diff</a>

</details>

### CODE REVIEW

**Review by coleenp — APPROVED** [review]
Looks good to have the largest if blocks possible.

File: src/hotspot/share/oops/method.cpp [hunk 1]
```diff
@@ -401,6 +401,7 @@ void Method::metaspace_pointers_do(MetaspaceClosure* it) {
   NOT_PRODUCT(it->push(&_name);)
 }
 
+#if INCLUDE_CDS
```
Comment (calvinccheung) [inline src/hotspot/share/oops/method.cpp:404]: Do you need to ifdef out the declaration in method.hpp?

**Review by calvinccheung — COMMENTED** [review]
Thanks for the cleanup. One comment below.

File: src/hotspot/share/oops/method.cpp (hunk 1 above)
Comment (iklam) [reply src/hotspot/share/oops/method.cpp:404]: Fixed. I also found out that `Method::restore_unshareable_info` also needed to be ifdef'ed out.

File: src/hotspot/share/oops/klass.hpp [hunk 2]
```diff
@@ -569,6 +572,7 @@ class Klass : public Metadata {
       return true;
     }
   }
+#endif
```
Comment (coleenp) [inline src/hotspot/share/oops/klass.hpp:575]: Can you add an // INCLUDE_CDS here?  If they're more than 10 lines apart, uncommented endifs can be really confusing.  This one is sort of borderline far away.  Thanks.

File: src/hotspot/share/oops/method.cpp (hunk 1 above)
Comment (calvinccheung) [reply src/hotspot/share/oops/method.cpp:404]: Updates look good.

**Review by calvinccheung — APPROVED** [review]

File: src/hotspot/share/oops/method.hpp [hunk 3]
```diff
@@ -134,7 +134,8 @@ class Method : public Metadata {
 
   virtual bool is_method() const { return true; }
 
-  void restore_unshareable_info(TRAPS);
+  void remove_unshareable_info()        NOT_CDS_RETURN;
+  void restore_unshareable_info(TRAPS)  NOT_CDS_RETURN;
```
Comment (dholmes-ora) [inline src/hotspot/share/oops/method.hpp:138]: Surely these should ifdef'd out not stubbed out? The callers of these should also be in INCLUDE_CDS blocks.

**Review by dholmes-ora — COMMENTED** [review]
This is a bit of a slippery slope. We have things like NOT_CDS_RETURN to avoid the need to use `ifdef INCLUDE_CDS` at the callsites. If you now ifdef the call sites you don't need NOT_CDS_RETURN. Conversely if you are still using NOT_CDS_RETURN then why didn't you ifdef the callsites?

File: src/hotspot/share/oops/method.hpp (hunk 3 above)
Comment (iklam) [reply src/hotspot/share/oops/method.hpp]: Fixed.

File: src/hotspot/share/oops/klass.hpp (hunk 2 above)
Comment (iklam) [reply src/hotspot/share/oops/klass.hpp]: Fixed.

**Review by dholmes-ora — APPROVED** [review]
Thanks for the updates.

**Review by calvinccheung — APPROVED** [review]

**Comment by iklam** [comment]
Thanks @calvinccheung @coleenp @dholmes-ora for the review!
/integrate

**Pull request closed without merging**